### ISSUE TITLE
Validate source passed to gl.shaderSource()

### ISSUE THREAD

**Issue by nox** [servo/servo#21754]
There are some illegal characters that must lead to a WebGL error being emitted.

**Comment by nox** [comment]
https://searchfox.org/mozilla-central/source/dom/canvas/WebGLShader.cpp#175-182

**Comment by sumit0190** [comment]
@nox I can take this up, if you think my understanding of the problem is correct:
1) Pre-process the source string to remove all sorts of comments, but preserve the line numbering.
2) Process the source string character-by-character and error out if an invalid character is found.

Not sure what's the most idiomatic way to do the first part. Also, there's an existing conformance test for this, right (I couldn't find it though)?

**Comment by jdm** [comment]
That is correct. The source is stored in a DOMString, which is just a wrapper around a String. I'm not sure if there's a more idiomatic way to implement the replacement than porting the C++ implementation to Rust.

**Comment by jdm** [comment]
@sumit0190 Are you still working on this? Have you made any progress?

**Comment by sumit0190** [comment]
Sorry, I got caught up in something and didn't find the time to publish the PR. I have my changes though, so I'll send out a request by tomorrow, if that's alright.

**Comment by jdm** [comment]
The patch in #21754 looked good but needed a few review comments addressed before it could merge.
